tss-esapi/src/tcti_ldr.rs

```diff
@@ -0,0 +1,617 @@
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::net::IpAddr;
+use std::path::PathBuf;
+use std::ptr::null_mut;
+use std::str::FromStr;
+
+const DEVICE: &str = "device";
+const MSSIM: &str = "mssim";
+const SWTPM: &str = "swtpm";
+const TABRMD: &str = "tabrmd";
+
+/// TCTI Context created via a TCTI Loader Library.
+/// Wrapper around the TSS2_TCTI_CONTEXT structure.
+#[derive(Debug)]
+#[allow(missing_copy_implementations)]
+pub struct TctiContext {
+    tcti_context: *mut tss_esapi_sys::TSS2_TCTI_CONTEXT,
+}
+
+impl TctiContext {
+    /// Allocate and initialize a new TctiContext structure.
+    pub fn initialize(name_conf: TctiNameConf) -> Result<Self> {
+        let mut tcti_context = null_mut();
+
+        let tcti_name_conf = CString::try_from(name_conf)?;
+
+        unsafe {
+            let ret =
+                tss_esapi_sys::Tss2_TctiLdr_Initialize(tcti_name_conf.as_ptr(), &mut tcti_context);
+            let ret = Error::from_tss_rc(ret);
+            if !ret.is_success() {
+                error!("Error when creating a TCTI context: {}", ret);
+                return Err(ret);
+            }
+        }
+        Ok(TctiContext { tcti_context })
+    }
+}
+
+impl Drop for TctiContext {
+    fn drop(&mut self) {
+        unsafe {
+            tss_esapi_sys::Tss2_TctiLdr_Finalize(&mut self.tcti_context);
+        }
+    }
+}
+
+/// Wrapper around the TSS2_TCTI_INFO structure.
+#[derive(Debug)]
+#[allow(missing_copy_implementations)]
+pub struct TctiInfo {
+    tcti_info: *mut tss_esapi_sys::TSS2_TCTI_INFO,
+}
+
+impl TctiInfo {
+    /// Query the TCTI loading mechanism
+    pub fn get_info(name_conf: TctiNameConf) -> Result<Self> {
+        let mut tcti_info = null_mut();
+
+        let tcti_name_conf = CString::try_from(name_conf)?;
+
+        unsafe {
+            let ret = tss_esapi_sys::Tss2_TctiLdr_GetInfo(tcti_name_conf.as_ptr(), &mut tcti_info);
+            let ret = Error::from_tss_rc(ret);
+            if !ret.is_success() {
+                error!("Error when getting the TCTI_INFO structure: {}", ret);
+                return Err(ret);
+            }
+        }
+        Ok(TctiInfo { tcti_info })
+    }
+
+    /// Get the version field
+    pub fn version(&self) -> u32 {
+        unsafe { (*(self.tcti_info)).version }
+    }
+
+    /// Get the name field
+    pub fn name(&self) -> &CStr {
+        unsafe { CStr::from_ptr((*(self.tcti_info)).name) }
+    }
+
+    /// Get the description field
+    pub fn description(&self) -> &CStr {
+        unsafe { CStr::from_ptr((*(self.tcti_info)).description) }
+    }
+
+    /// Get the config_help field
+    pub fn config_help(&self) -> &CStr {
+        unsafe { CStr::from_ptr((*(self.tcti_info)).config_help) }
+    }
+}
+
+impl Drop for TctiInfo {
+    fn drop(&mut self) {
+        unsafe {
+            tss_esapi_sys::Tss2_TctiLdr_FreeInfo(&mut self.tcti_info);
+        }
+    }
+}
+
+/// Placeholder TCTI types that can be used when initialising a `Context` to determine which
```

Comment:
From down below everything is copied but changing `Tcti` to `TctiNameConf` to match the spec.